p4-16/psa/examples/psa-example-parser-checksum.p4

```diff
@@ -0,0 +1,232 @@
+// type cannot be directly put into packet headers or used as an index
+// into counter, meter, or register arrays, so such a conversion
+// 'function' is necessary to get a value that can be used for these
+// purposes.
+
+action convert_error_to_idx (in ParserError_t err,
+                             out ErrorIndex_t idx)
+{
+    if (err == error.NoError) {
+        idx = 1;
+    } else if (err == error.PacketTooShort) {
+        idx = 2;
+    } else if (err == error.NoMatch) {
+        idx = 3;
+    } else if (err == error.StackOutOfBounds) {
+        idx = 4;
+    } else if (err == error.HeaderTooShort) {
+        idx = 5;
+    } else if (err == error.ParserTimeout) {
+        idx = 6;
+    } else if (err == error.BadIPv4HeaderChecksum) {
+        idx = 7;
+    } else if (err == error.UnhandledIPv4Options) {
+        idx = 8;
+    } else {
+        // Unknown error value
+        idx = 0;
+    }
+}
+
+parser IngressParserImpl(packet_in buffer,
+                         out headers parsed_hdr,
+                         inout metadata user_meta,
+                         in psa_ingress_parser_input_metadata_t istd,
+                         out psa_parser_output_metadata_t ostd)
+{
+    Checksum<bit<16>>(HashAlgorithm.ones_complement16) ck;
+    state start {
+        buffer.extract(parsed_hdr.ethernet);
+        transition select(parsed_hdr.ethernet.etherType) {
+            0x0800: parse_ipv4;
+            default: accept;
+        }
+    }
+    state parse_ipv4 {
+        buffer.extract(parsed_hdr.ipv4);
+        // TBD: It would be good to enhance this example to
+        // demonstrate checking of IPv4 header checksums for IPv4
+        // headers with options, but this example does not handle such
+        // packets.
+        verify(parsed_hdr.ipv4.ihl == 5, error.UnhandledIPv4Options);
+        // TBD: Update the code below for checking the IPv4 header
+        // checksum with whatever the API we decide upon for a PSA
+        // checksum extern.
+        ck.clear();
+        ck.update({ parsed_hdr.ipv4.version,
+                parsed_hdr.ipv4.ihl,
+                parsed_hdr.ipv4.diffserv,
+                parsed_hdr.ipv4.totalLen,
+                parsed_hdr.ipv4.identification,
+                parsed_hdr.ipv4.flags,
+                parsed_hdr.ipv4.fragOffset,
+                parsed_hdr.ipv4.ttl,
+                parsed_hdr.ipv4.protocol,
+                parsed_hdr.ipv4.hdrChecksum,
+                parsed_hdr.ipv4.srcAddr,
+                parsed_hdr.ipv4.dstAddr });
+        // The verify statement below will cause the parser to enter
+        // the reject state, and thus terminate parsing immediately,
+        // if the IPv4 header checksum is wrong.  It will also record
+        // the error error.BadIPv4HeaderChecksum, which will be
+        // available in a metadata field in the ingress control block.
+        verify(ck.get() == 0, error.BadIPv4HeaderChecksum);
```

Comment:
Thanks, this is pretty much what I had in mind. However, there is a little difference with the code I've written and I am just wondering if the compiler will be able to do the right thing in both cases. I hope it is possible.

My code looked like this:

```
ck.update({ parsed_hdr.ipv4.version,
                parsed_hdr.ipv4.ihl,
                parsed_hdr.ipv4.diffserv,
                parsed_hdr.ipv4.totalLen,
                parsed_hdr.ipv4.identification,
                parsed_hdr.ipv4.flags,
                parsed_hdr.ipv4.fragOffset,
                parsed_hdr.ipv4.ttl,
                parsed_hdr.ipv4.protocol,
                // parsed_hdr.ipv4.hdrChecksum,
                parsed_hdr.ipv4.srcAddr,
                parsed_hdr.ipv4.dstAddr });
        // The verify statement below will cause the parser to enter
        // the reject state, and thus terminate parsing immediately,
        // if the IPv4 header checksum is wrong.  It will also record
        // the error error.BadIPv4HeaderChecksum, which will be
        // available in a metadata field in the ingress control block.
        verify(ck.get() == parsed_hdr.ipv4.hdrChecksum, error.BadIPv4HeaderChecksum);
```

Reply:
I would expect that a checksum extern API could be implemented to work in the way that you show, and can change the example to demonstrate it that way.

Reply:
Example code changed to leave ipv4.hdrChecksum out of checksum calculation, instead using it to compare output of checksum calculation against that value.